ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/GenerateSignatureFileAction.fs

```diff
@@ -0,0 +1,100 @@
+namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Features.ContextActions
+
+open System.IO
+open JetBrains.ReSharper.Feature.Services.ContextActions
+open JetBrains.ReSharper.Plugins.FSharp.Psi
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Intentions
+open type JetBrains.ReSharper.Psi.PsiSourceFileExtensions
+open JetBrains.ReSharper.Plugins.FSharp
+open  JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open JetBrains.ReSharper.Psi
+open JetBrains.ReSharper.Psi.ExtensionsAPI.Tree
+
+type TopLevelModuleOrNamespace =
+    {
+        IsModule: bool
+        Name: string
+        NestedModules: NestedModule list
+    }
+
+and NestedModule =
+    {
+        Name: string
+        NestedModules: NestedModule list
+    }
+
+[<ContextAction(Group = "F#", Name = "Generate signature file for current file", Priority = 1s,
+                Description = "Generate signature file for current file.")>]
+type GenerateSignatureFileAction(dataProvider: FSharpContextActionDataProvider) =
+    inherit FSharpContextActionBase(dataProvider)
+    
+    let mkSignatureFile (fsharpFile: IFSharpFile) : IFSharpFile =
+        let factory : IFSharpElementFactory = fsharpFile.CreateElementFactory()
+        let signatureFile : IFSharpFile = factory.CreateEmptyFile()
+        
+        let rec processModuleMembers (parent: IFSharpTreeNode) (members: IModuleMember seq) =
+            for m in members do
+                match m with
+                | :? INestedModuleDeclaration as nmd ->
+                    let nestedModuleNode = factory.CreateNestedModule(nmd.NameIdentifier.Name)
+                    // TODO: if the nested module has nested modules, clear the content (`begin end`) and process them.
+                    ModificationUtil.AddChild(parent, nestedModuleNode) |> ignore
+                | _ -> ()
+        
+        for decl in fsharpFile.ModuleDeclarations do
```

Comment:
@auduchinok I know you mentioned using the `TreeNodeVisitor` here, but I don't quite see how to easily use that here. Any suggestions?

Reply:
It's OK to do it this way too, maybe it'll be even easier at the end